File: src/main/java/me/taubsie/dungeonhub/application/classes/PlayerInformation.java
package me.taubsie.dungeonhub.application.classes;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import me.taubsie.dungeonhub.application.connection.HypixelConnection;
import me.taubsie.dungeonhub.application.connection.MojangConnection;
import me.taubsie.dungeonhub.common.model.discord_user.DiscordUserModel;
import org.javacord.api.entity.user.User;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class PlayerInformation {
    private User user;
    private DiscordUserModel discordUserModel;

    public Map<String, Supplier<String>> getReplacements() {
        Map<String, Supplier<String>> replacements = new HashMap<>();

        replacements.put("discord.name", () -> user.getName());
        //TODO use display name as soon as javacord catches on
        replacements.put("discord.displayname", () -> user.getName());
        replacements.put("minecraft.name", () -> MojangConnection.getInstance()
                .getNameByUUID(discordUserModel.getMinecraftId()));
        replacements.put("skyblock.catacombs.level", () -> String.valueOf(HypixelConnection.getInstance()
                .getCataLevelByUUID(discordUserModel.getMinecraftId())));
        replacements.put("skyblock.level", () -> String.valueOf(HypixelConnection.getInstance().getSkyblockLevelByUUID(discordUserModel.getMinecraftId())));

        return replacements;
    }
}

File: src/main/java/me/taubsie/dungeonhub/kord/application/misc/PlayerInformation.kt
package me.taubsie.dungeonhub.kord.application.misc

import dev.kord.core.entity.User
import dev.kord.core.entity.effectiveName
import lombok.AllArgsConstructor
import lombok.Getter
import lombok.NoArgsConstructor
import lombok.Setter
import me.taubsie.dungeonhub.application.connection.MojangConnection
import me.taubsie.dungeonhub.common.model.discord_user.DiscordUserModel
import me.taubsie.dungeonhub.kord.application.connection.HypixelConnection.getCataLevelByUUID
import me.taubsie.dungeonhub.kord.application.connection.HypixelConnection.getSkyblockLevelByUUID

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
class PlayerInformation(private val user: User, private val discordUserModel: DiscordUserModel) {
    val replacements: Map<String, () -> String>
        get() {
            val replacements: MutableMap<String, () -> String> = HashMap()

            replacements["discord.name"] = { user.username }
            replacements["discord.displayname"] = { user.effectiveName }
            replacements["minecraft.name"] = {
                MojangConnection.getInstance()
                    .getNameByUUID(discordUserModel.minecraftId)
            }
            replacements["skyblock.catacombs.level"] =
                { getCataLevelByUUID(discordUserModel.minecraftId).toString() }
            replacements["skyblock.level"] = {
                getSkyblockLevelByUUID(
                    discordUserModel.minecraftId
                ).stream().mapToObj { i: Int -> java.lang.String.valueOf(i) }.findAny().orElse("?")
            }

            return replacements
        }
}